Source/Core/Core/DSP/Interpreter/DSPIntTables.cpp

```diff
@@ -0,0 +1,391 @@
+  {0x8b00, 0xff00, srbith},
+  {0x8c00, 0xff00, srbith},
+  {0x8d00, 0xff00, srbith},
+  {0x8e00, 0xff00, srbith},
+  {0x8f00, 0xff00, srbith},
+
+  // 9
+  {0x9000, 0xf700, mul},
+  {0x9100, 0xf700, asr16},
+  {0x9200, 0xf600, mulmvz},
+  {0x9400, 0xf600, mulac},
+  {0x9600, 0xf600, mulmv},
+
+  // A-B
+  {0xa000, 0xe700, mulx},
+  {0xa100, 0xf700, abs},
+  {0xa200, 0xe600, mulxmvz},
+  {0xa400, 0xe600, mulxac},
+  {0xa600, 0xe600, mulxmv},
+  {0xb100, 0xf700, tst},
+
+  // C-D
+  {0xc000, 0xe700, mulc},
+  {0xc100, 0xe700, cmpar},
+  {0xc200, 0xe600, mulcmvz},
+  {0xc400, 0xe600, mulcac},
+  {0xc600, 0xe600, mulcmv},
+
+  // E
+  {0xe000, 0xfc00, maddx},
+  {0xe400, 0xfc00, msubx},
+  {0xe800, 0xfc00, maddc},
+  {0xec00, 0xfc00, msubc},
+
+  // F
+  {0xf000, 0xfe00, lsl16},
+  {0xf200, 0xfe00, madd},
+  {0xf400, 0xfe00, lsr16},
+  {0xf600, 0xfe00, msub},
+  {0xf800, 0xfc00, addpaxz},
+  {0xfc00, 0xfe00, clrl},
+  {0xfe00, 0xfe00, movpz},
+}};
+
+constexpr std::array<InterpreterOpInfo, 25> s_opcodes_ext
+{{
+  {0x0000, 0x00fc, Ext::nop},
+
+  {0x0004, 0x00fc, Ext::dr},
+  {0x0008, 0x00fc, Ext::ir},
+  {0x000c, 0x00fc, Ext::nr},
+  {0x0010, 0x00f0, Ext::mv},
+
+  {0x0020, 0x00e4, Ext::s},
+  {0x0024, 0x00e4, Ext::sn},
+
+  {0x0040, 0x00c4, Ext::l},
+  {0x0044, 0x00c4, Ext::ln},
+
+  {0x0080, 0x00ce, Ext::ls},
+  {0x0082, 0x00ce, Ext::sl},
+  {0x0084, 0x00ce, Ext::lsn},
+  {0x0086, 0x00ce, Ext::sln},
+  {0x0088, 0x00ce, Ext::lsm},
+  {0x008a, 0x00ce, Ext::slm},
+  {0x008c, 0x00ce, Ext::lsnm},
+  {0x008e, 0x00ce, Ext::slnm},
+
+  {0x00c3, 0x00cf, Ext::ldax},
+  {0x00c7, 0x00cf, Ext::ldaxn},
+  {0x00cb, 0x00cf, Ext::ldaxm},
+  {0x00cf, 0x00cf, Ext::ldaxnm},
+
+  {0x00c0, 0x00cc, Ext::ld},
+  {0x00c4, 0x00cc, Ext::ldn},
+  {0x00c8, 0x00cc, Ext::ldm},
+  {0x00cc, 0x00cc, Ext::ldnm},
+}};
+// clang-format on
+
+namespace
+{
+std::array<InterpreterFunction, 65536> s_op_table;
+std::array<InterpreterFunction, 256> s_ext_op_table;
+bool s_tables_initialized = false;
+
+template <size_t N>
+auto FindByOpcode(UDSPInstruction opcode, const std::array<InterpreterOpInfo, N>& data)
+{
+  return std::find_if(data.cbegin(), data.cend(),
+                      [opcode](const auto& info) { return (opcode & info.mask) == info.opcode; });
+}
+}  // Anonymous namespace
+
+InterpreterFunction GetOp(UDSPInstruction inst)
+{
+  return s_op_table[inst];
+}
+
+InterpreterFunction GetExtOp(UDSPInstruction inst)
+{
+  const bool has_seven_bit_extension = (inst >> 12) == 0x3;
+
+  if (has_seven_bit_extension)
+    return s_ext_op_table[inst & 0x7F];
+
+  return s_ext_op_table[inst & 0xFF];
+}
+
+void InitInstructionTables()
+{
+  if (s_tables_initialized)
+    return;
+
+  // ext op table
+  for (size_t i = 0; i < s_ext_op_table.size(); i++)
+  {
+    s_ext_op_table[i] = nop;
+
+    const auto iter = FindByOpcode(static_cast<UDSPInstruction>(i), s_opcodes_ext);
+    if (iter == s_opcodes_ext.cend())
+      continue;
+
+    if (s_ext_op_table[i] == nop)
```

Comment:
This condition is always true, isn't it? You set s_ext_op_table[i] = nop; 6 lines before.

Reply:
Not if the find hits the end of the table.

Reply:
They this condition would still be true, but not evaluated.

Reply:
Ah, true, I'll remove this.